File: lib/xd/_webview.ts
import { WebViewAttributes,WindowTrait,Size } from './mod.ts';
import * as lib from '../../bindings/bindings.ts';
import { Color,$rgba } from "../screencapture/color.ts";
import { $unimplemented,Option } from '../mod.ts';

export abstract class WebView {
  protected webviewAttrs: WebViewAttributes={};

  protected _window=new class extends WindowTrait {
    protected windowAttrs={};
    public _addrs=new BigUint64Array(2);

    protected get _window(): bigint {
      return this._addrs[0];
    }
  };
  
  private get _webview() {
    return this._window._addrs[1];
  }
  
  public clearAllBrowsingData() {
    this._webview && lib.clear_all_browsing_data(this._webview);
  }

  public eval(js: string) {
    this._webview && lib.eval_script(this._webview,js);
  }

  public webViewInnerSize(): Size {
    return this._webview?JSON.parse(lib.webview_inner_size(this._webview)):WindowTrait.defaultSize;
  }

  public loadUrl(url: string|URL) {
    this._webview?lib.load_url(this._webview,url.toString()):this.webviewAttrs.url=url;
  }
  
  public lordUrlWithHeaders(url: string|URL,headers: Record<string,string>) {
    this._webview && lib.load_url_with_headers(this._webview,url.toString(),JSON.stringify(headers));
  }
  
  public print() {
    this._webview && lib.webview_print(this._webview);
  }

  public setBackgroundColor(color: Color) {
    const rgba=$rgba(color);
    this._webview?lib.set_background_color(this._webview,rgba.r,rgba.g,rgba.b,rgba.a):this.webviewAttrs.backgroundColor=rgba;
  }

  public url() {
    return new Option((this._webview && lib.url(this._webview))||null);
  }

  public window() {
    $unimplemented();
  }

  public zoom(scaleFactor: number) {
    this._webview && lib.zoom(this._webview,scaleFactor);
  }
}

File: src/xd/webview.rs

use xd_macro::method;
use wry::webview::WebView;

use crate::{
  Size,
  to_header_map,
  exception::Exception
};

use deno_bindgen::{
  deno_bindgen,
  serde_json as json
};

#[method]
pub fn clear_all_browsing_data(this: &WebView) {
  this.clear_all_browsing_data().unwrap_or_throw()
}

#[method]
pub fn eval_script(this: &WebView,js: &str) {
  this.evaluate_script(js).unwrap_or_throw()
}

#[method]
pub fn webview_inner_size(this: &WebView)-> String {
  json::to_string(&Size::from(this.inner_size())).unwrap_or_throw()
}

#[method]
pub fn load_url(this: &WebView,url: &str) {
  this.load_url(url)
}

#[method]
pub fn load_url_with_headers(this: &WebView,url: &str,headers: &str) {
  this.load_url_with_headers(url,to_header_map(json::from_str(headers).unwrap_or_throw()))
}

#[method]
pub fn webview_print(this: &WebView) {
  this.print().unwrap_or_throw()
}

#[method]
pub fn set_background_color(this: &WebView,r: u8,g: u8,b: u8,a: u8) {
  this.set_background_color((r,g,b,a)).unwrap_or_throw()
}

#[method]
pub fn url(this: &WebView)-> String {
  this.url().to_string()
}

#[method]
pub fn window(this: &WebView)-> usize {
  this.window() as *const _ as _
}

#[method]
pub fn zoom(this: &WebView,zoom: f64) {
  this.zoom(zoom)
}

